### PR TITLE
[OTE-776] Implement GetAllRevshares

### PR DESCRIPTION
**Pull request by affanv14**
### Changelist
Implement functions to get all revshares for all market types

### Test Plan
Adds unit tests

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new interface and struct for processing perpetual fills, improving data organization and encapsulation.
	- Enhanced revenue share calculations with new methods for retrieving and aggregating shares from various sources.
	- Added a constant for maximum referee transaction volume over 30 days to ensure consistency in financial metrics.
	- Implemented new error handling for scenarios where total fees shared exceed net fees.

- **Bug Fixes**
	- Improved error handling in revenue share calculations to manage potential issues effectively.

- **Tests**
	- Added tests to validate revenue share retrieval under both valid and invalid scenarios, enhancing the reliability of the revenue sharing logic.

### CODE REVIEW

File: protocol/x/revshare/keeper/revshare.go [hunk 1]
```diff
@@ -142,3 +146,127 @@ func (k Keeper) ValidateRevShareSafety(
 	totalRevSharePpm := totalUnconditionalRevSharePpm + totalMarketMapperRevSharePpm + highestTierRevSharePpm
 	return totalRevSharePpm < lib.OneMillion
 }
+
+func (k Keeper) GetAllRevShares(
+	ctx sdk.Context,
+	fill clobtypes.FillForProcess,
+	monthlyRollingTakerVolumeQuantums uint64,
```
Comment (affanv14) [inline protocol/x/revshare/keeper/revshare.go:153]: @teddyding Added this extra field here instead of adding stakingkeeper to reduce dependencies. Also makes it easier to test

File: protocol/x/clob/types/types.go [hunk 2]
```diff
@@ -1 +1,63 @@
 package types
+
+import "math/big"
```
Comment (affanv14) [inline protocol/x/clob/types/types.go:3]: Didnt put this in shared types. Since clob types and revshare types dont depend on each other - i didnt think this was necessary

File: protocol/x/revshare/keeper/revshare.go (hunk 1 above)
Comment (teddyding) [reply protocol/x/revshare/keeper/revshare.go:153]: Let's just add it to `FillForProcess`? Seems cleaner

Comment (teddyding) [reply protocol/x/revshare/keeper/revshare.go:153]: Also let's document that this rolling volume is the stats at the end of last block - as in, if trader has multiple fills in the block currently being processed, this trailing volume doesn't update intra block per fill.

File: protocol/x/revshare/keeper/revshare.go [hunk 3]
```diff
@@ -142,3 +146,127 @@ func (k Keeper) ValidateRevShareSafety(
...
+	netFees *big.Int,
+) ([]types.RevShare, *big.Int, error) {
+	revShares := []types.RevShare{}
+	feesShared := big.NewInt(0)
+	marketMapperRevshareAddress, revenueSharePpm, err := k.GetMarketMapperRevenueShareForMarket(ctx, marketId)
+	if err != nil {
+		return revShares, feesShared, err
+	}
+	if revenueSharePpm == 0 {
+		return revShares, feesShared, nil
```
Comment (teddyding) [inline protocol/x/revshare/keeper/revshare.go:259]: Nit: explicitly return `nil` list and `0` fee shared

File: protocol/x/clob/types/types.go [hunk 4]
```diff
@@ -1 +1,63 @@
...
+
+type FillForProcess interface {
+	TakerAddr() string
+	TakerFeeQuoteQuantums() *big.Int
+	MakerAddr() string
+	MakerFeeQuoteQuantums() *big.Int
+	FillQuoteQuantums() *big.Int
+	PerpetualId() uint32
```
Comment (teddyding) [inline protocol/x/clob/types/types.go:11]: The original intention of having this interface is so it applies to both perpetual and spot in the future. If we calling direct calling this `PerpetualId` at interface level, might as we just delete `FillForProcess` and use `PerpetualFillForProcess` struct directly

File: protocol/x/revshare/keeper/revshare.go [hunk 5]
```diff
@@ -142,3 +146,123 @@ func (k Keeper) ValidateRevShareSafety(
...
+	unconditionalRevShareConfig, err := k.GetUnconditionalRevShareConfigParams(ctx)
+	if err != nil {
+		return nil, big.NewInt(0), err
+	}
+	for _, revShare := range unconditionalRevShareConfig.Configs {
+		feeShared := lib.BigMulPpm(netFees, lib.BigU(revShare.SharePpm), false)
+		totalFeesShared.Add(totalFeesShared, feeShared)
+		revShare := types.RevShare{
+			Recipient:         revShare.Address,
+			RevShareFeeSource: types.REV_SHARE_FEE_SOURCE_NET_FEE,
```
Comment (teddyding) [inline protocol/x/revshare/keeper/revshare.go:234]: On second thought, do we need this in any downstream logic?

File: protocol/x/revshare/keeper/revshare_test.go [hunk 6]
```diff
@@ -261,3 +266,320 @@ func TestValidateRevShareSafety(t *testing.T) {
...
+		name                              string
+		revenueSharePpmNetFees            uint32
+		revenueSharePpmTakerFees          uint32
+		expectedRevShares                 int
+		monthlyRollingTakerVolumeQuantums uint64
+		setup                             func(tApp *testapp.TestApp, ctx sdk.Context,
+			keeper *keeper.Keeper, affiliatesKeeper *affiliateskeeper.Keeper)
+	}{
+		{
+			name:                              "Valid revenue share",
```
Comment (teddyding) [inline protocol/x/revshare/keeper/revshare_test.go:281]: Nit: could we rename to more descriptive test case. E.g. `affiliate + unconditional + marketmapper`

File: protocol/x/revshare/keeper/revshare_test.go [hunk 7]
```diff
@@ -261,3 +266,320 @@ func TestValidateRevShareSafety(t *testing.T) {
 		})
 	}
 }
+
+func TestKeeper_GetAllRevShares_Valid(t *testing.T) {
+	tests := []struct {
+		name                              string
+		revenueSharePpmNetFees            uint32
+		revenueSharePpmTakerFees          uint32
+		expectedRevShares                 int
```
Comment (teddyding) [inline protocol/x/revshare/keeper/revshare_test.go:275]: Should have more granular expected value than just # of rev shares. e.g. `[]types.RevShare`

File: protocol/x/revshare/keeper/revshare.go (hunk 5 above)
Comment (affanv14) [reply protocol/x/revshare/keeper/revshare.go:234]: Dont we need revshare Fee source in [trading rewards formula](https://www.notion.so/dydx/RevShare-Generalization-6d89820112c94dfeaefc5a5ace052cf4?pvs=4#893602cca8b945548006414c877d2ed8)?

Comment (affanv14) [reply protocol/x/revshare/keeper/revshare.go:234]: And revshare type to retrieve affiliate revshares later for sending indexer events

Comment (teddyding) [reply protocol/x/revshare/keeper/revshare.go:231]: > Dont we need revshare Fee source in [trading rewards formula](https://www.notion.so/dydx/RevShare-Generalization-6d89820112c94dfeaefc5a5ace052cf4?pvs=4#893602cca8b945548006414c877d2ed8)?

Ah yes we need so that we can separate reward logic for maker and taker fee

File: protocol/x/revshare/keeper/revshare_test.go [hunk 8]
```diff
@@ -268,26 +267,52 @@ func TestValidateRevShareSafety(t *testing.T) {
...
+					QuoteQuantums:     big.NewInt(3_600_000),
+				},
+				{
+					Recipient:         constants.AliceAccAddress.String(),
+					RevShareFeeSource: types.REV_SHARE_FEE_SOURCE_NET_FEE,
+					RevShareType:      types.REV_SHARE_TYPE_MARKET_MAPPER,
+					QuoteQuantums:     big.NewInt(1_200_000),
+				},
 			},
+			fill: clobtypes.CreatePerpetualFillForProcess(
```
Comment (teddyding) [inline protocol/x/revshare/keeper/revshare_test.go:307]: This function actually makes the code less readable, could we just replace this with a struct literal whereever it's used? For example it's hard for reader to tell what these fields are

File: protocol/x/revshare/keeper/revshare_test.go [hunk 9]
```diff
@@ -353,15 +400,31 @@ func TestKeeper_GetAllRevShares_Valid(t *testing.T) {
...
+				constants.AliceAccAddress.String(),
+				big.NewInt(10_000_000),
+				constants.BobAccAddress.String(),
+				big.NewInt(2_000_000),
+				big.NewInt(100000),
```
Comment (teddyding) [inline protocol/x/revshare/keeper/revshare_test.go:424]: Fees are usually in the magnitude of `~0.01%` of fill quote quantums, so let's use more realistic values here

File: protocol/x/revshare/keeper/revshare_test.go [hunk 10]
```diff
@@ -317,13 +342,35 @@ func TestKeeper_GetAllRevShares_Valid(t *testing.T) {
...
 		},
 		{
-			name:                                  "Valid revenue share with 30d volume greater than max 30d referral volume",
-			expectedRevShares:                     3,
-			monthlyRollingTakerVolumeQuantums:     types.Max30dRefereeVolumeQuantums + 1,
-			expectedTakerFeeRevSharesPpmByAddress: map[string]uint32{},
-			expectedNetFeeRevSharesPpmByAddress: map[string]uint32{
-				constants.AliceAccAddress.String(): 400_000, // 40% of net fees
-				constants.BobAccAddress.String():   200_000, // 20% of net fees
+			name: "Valid revenue share with 30d volume greater than max 30d referral volume",
```
Comment (teddyding) [inline protocol/x/revshare/keeper/revshare_test.go:345]: Can you add a test where maker fee is negative and taker fee is positive, and check that market mapper and unconditonal rev share are based net fee

File: protocol/x/revshare/keeper/revshare_test.go (hunk 8 above)
Comment (affanv14) [reply protocol/x/revshare/keeper/revshare_test.go:307]: If we want to get rid of the function and update the struct directly, we have to rename the fields to be capitalized(same as the interface - causing errors) 
two options here:
1. Remove interface entirely and stick with structs
2. Change the anme of underlying stuct vars(like PerpetualMarketAddr or MarketAddr_) which might be a little more confusing

Comment (affanv14) [reply protocol/x/revshare/keeper/revshare_test.go:307]: im leaning 1

**Review by teddyding — APPROVED** [review]
